### PR TITLE
Silu op for Sharded layout

### PR DESCRIPTION
**Pull request by shwetankTT**
Performance details for Sharded vs not sharded data.
https://docs.google.com/spreadsheets/d/1yM6-z1OgCH69x2ijqyjI0Pm0dRQnACxLSshZsvzdfQM/edit?usp=sharing'

### CODE REVIEW

File: tt_eager/tt_dnn/op_library/eltwise_unary/eltwise_unary_op.cpp [hunk 1]
```diff
@@ -257,18 +257,58 @@ void EltwiseUnary::validate(const std::vector<Tensor> &input_tensors) const {
...
 
 std::vector<Tensor> EltwiseUnary::create_output_tensors(const std::vector<Tensor> &input_tensors) const {
     const auto& input_tensor = input_tensors.at(0);
+    auto input_shard_spec = input_tensor.memory_config().shard_spec.value();
+     if(this->output_mem_config.is_sharded()){
+        if(input_tensor.memory_config().memory_layout == TensorMemoryLayout::HEIGHT_SHARDED){
+            auto ncores = input_shard_spec.num_cores();
+            Shape output_shape = compute_output_shapes(input_tensors).at(0);
+            array<uint32_t, 2> output_shard_shape = {div_up(output_shape[0] * output_shape[1] * output_shape[2], ncores), output_shape[-1]};
+            std::cout << __FILE__ << " " << __LINE__ << " " << output_shard_shape << std::endl;
```
Comment (nsmithtt) [inline tt_eager/tt_dnn/op_library/eltwise_unary/eltwise_unary_op.cpp:289]: stray print

File: tt_eager/tt_dnn/kernels/dataflow/reader_unary_sharded.cpp [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+#include "debug/dprint.h"
+
+void kernel_main() {
+    uint32_t num_tiles_per_core = get_arg_val<uint32_t>(0);
+    constexpr uint32_t cb_id_in0 = get_compile_time_arg_val(0);
+
+    constexpr uint32_t onetile = 1;
+    for (uint32_t i = 0; i < num_tiles_per_core; ++ i) {
+        cb_push_back(cb_id_in0, onetile);
+        DPRINT << "reader Unary_done " << ENDL();
```
Comment (mywoodstock) [inline tt_eager/tt_dnn/kernels/dataflow/reader_unary_sharded.cpp:17]: Remove dprint and the include

File: tt_eager/tt_dnn/op_library/eltwise_unary/eltwise_unary_op.cpp [hunk 3]
```diff
@@ -257,18 +257,58 @@ void EltwiseUnary::validate(const std::vector<Tensor> &input_tensors) const {
     const auto& input_tensor_a = input_tensors.at(0);
     TT_FATAL(input_tensor_a.storage_type() == StorageType::DEVICE, "Operands to eltwise unary need to be on device!");
     TT_FATAL(input_tensor_a.buffer() != nullptr , "Operands to eltwise unary need to be allocated in buffers on device!");
-    TT_FATAL((input_tensor_a.get_layout() == Layout::TILE), "Inputs to eltwise unary must be tilized");
+    /*TT_FATAL((input_tensor_a.get_layout() == Layout::TILE), "Inputs to eltwise unary must be tilized");
     TT_FATAL(input_tensor_a.memory_config().memory_layout == TensorMemoryLayout::INTERLEAVED, "Eltwise unary does not currently support sharding");
-    TT_FATAL(this->output_mem_config.memory_layout == TensorMemoryLayout::INTERLEAVED, "Eltwise unary does not currently support sharding");
+    TT_FATAL(this->output_mem_config.memory_layout == TensorMemoryLayout::INTERLEAVED, "Eltwise unary does not currently support sharding");*/
```
Comment (mywoodstock) [inline tt_eager/tt_dnn/op_library/eltwise_unary/eltwise_unary_op.cpp:262]: delete?

File: tt_eager/tt_dnn/op_library/eltwise_unary/multi_core/eltwise_unary_op_multi_core.cpp [hunk 4]
```diff
@@ -17,7 +17,185 @@ namespace tt {
...
+        const std::vector<Buffer*>& input_buffers,
+        const std::vector<Buffer*>& output_buffers
+    ) {
+
+        auto src_buffer = input_buffers.at(0);
+        auto dst_buffer = output_buffers.at(0);
+        for (uint32_t core = 0; core < ncores_nhw; core++){
+            CoreCoord core_coord(core % ncores_x, core / ncores_x); //
+            {
+
```
Comment (mywoodstock) [inline tt_eager/tt_dnn/op_library/eltwise_unary/multi_core/eltwise_unary_op_multi_core.cpp:186]: something missing?

File: tests/ttnn/unit_tests/operations/test_silu.py [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+                    ttnn.experimental.tensor.CoreCoord(0, 0),
+                    ttnn.experimental.tensor.CoreCoord(ncores_w - 1, ncores_h - 1),
+                )
+            }
+        )
+    else:
+        raise ValueError("Invalid ncores")
+
+
+@skip_for_wormhole_b0()
```
[the remark this answers is not in the document]
Comment (shwetankTT) [reply tests/ttnn/unit_tests/operations/test_silu.py:68]: > we should have this on WH too :)

Will test on WH too.

File: ttnn/ttnn/operations/unary.py [hunk 6]
```diff
@@ -50,7 +50,7 @@ def _unary_validate_input_tensors(operation_name, input_tensor, *args, **kwargs)
             input_tensor,
             ranks=(2, 3, 4),
             dtypes=(ttnn.bfloat16, ttnn.bfloat8_b),
-            layouts=(ttnn.TILE_LAYOUT,),
+            layouts=(ttnn.TILE_LAYOUT, ttnn.ROW_MAJOR_LAYOUT),
```
Comment (eyonland) [inline ttnn/ttnn/operations/unary.py:53]: Is it necessary to allow ttnn.ROW_MAJOR_LAYOUT?

**Review by eyonland — CHANGES_REQUESTED** [review]

File: ttnn/ttnn/operations/unary.py (hunk 6 above)
Comment (shwetankTT) [reply ttnn/ttnn/operations/unary.py:53]: > Is it necessary to allow ttnn.ROW_MAJOR_LAYOUT?

hey @eyonland,
Any reason why we should not allow ROW_MAJOR_LAYOUR? Perf reason?

Comment (shwetankTT) [reply ttnn/ttnn/operations/unary.py:53]: This is the requirement for this issue. https://github.com/tenstorrent-metal/tt-metal/issues/6184

File: tests/ttnn/unit_tests/operations/test_silu.py [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+from models.utility_functions import skip_for_wormhole_b0
+from tt_lib.utils import (
+    _nearest_y,
+)
+
+TILE_WIDTH = 32
+
+
+def get_shard_grid_from_num_cores(ncores: Union[int, Tuple[int, int]]) -> ttnn.experimental.tensor.CoreRangeSet:
+    max_grid_size = (9, 12)  ## (y, x)
```
Comment (tt-nshanker) [inline tests/ttnn/unit_tests/operations/test_silu.py:26]: max grid for WH is different. Please use an api to get max grid given a device object. There is definitely one in ttlib and I am not sure if it has been ported to TTNN. @tapspatel do you know the ttnn api to get max grid size from device?

File: tests/ttnn/unit_tests/operations/test_silu.py [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+    ## input shard
+    if shard_strategy == ttnn.ShardStrategy.BLOCK:
+        input_2d_height_padded = _nearest_y(input_2d_height, shard_grid[0] * 32)
+        shard_height = math.ceil(input_2d_height_padded / shard_grid[0])
+        shard_width = math.ceil(input_2d_width / shard_grid[1])
+        # shard_orientation = ttnn.ShardOrientation.COLUMN_MAJOR
+        shard_orientation = ttnn.experimental.tensor.ShardOrientation.COL_MAJOR
+        tensor_memory_layout = ttnn.types.TensorMemoryLayout.BLOCK_SHARDED
+    elif shard_strategy == ttnn.ShardStrategy.HEIGHT:
+        shard_height = math.ceil(input_2d_height / ncores)
```
Comment (tt-nshanker) [inline tests/ttnn/unit_tests/operations/test_silu.py:122]: We need the height padding to 32 for height sharding too. Basically what you did for block sharding.

File: tests/ttnn/unit_tests/operations/test_silu.py [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+        shard_orientation = ttnn.experimental.tensor.ShardOrientation.ROW_MAJOR
+        tensor_memory_layout = ttnn.types.TensorMemoryLayout.HEIGHT_SHARDED
+
+    core_grid = ttnn.CoreGrid(y=shard_grid[1], x=shard_grid[0])
+    shard_grid = get_shard_grid_from_num_cores(ncores)
+    logger.debug(f"shard_grid={shard_grid}")
+    logger.debug(f"grid_size={core_grid}")
+    logger.debug(f"input_shard_height={shard_height}, input_shard_width={shard_width}")
+
+    shard_spec = ttnn.experimental.tensor.ShardSpec(shard_grid, (shard_height, shard_width), shard_orientation, False)
```
Comment (tt-nshanker) [inline tests/ttnn/unit_tests/operations/test_silu.py:134]: Please add assert that shard height and width are divisible by 32.

File: tests/ttnn/unit_tests/operations/test_silu.py (hunk 7 above)
Comment (tapspatel) [reply tests/ttnn/unit_tests/operations/test_silu.py:26]: you can get device from the tensor via `input_tensor.device()`
pass that into your function
query via `max_grid_size = (8, 8) if device.arch() == ttl.device.Arch.WORMHOLE_B0 else (9, 12)  ## (y, x)`
check `ttnn/ttnn/operations/others.py` function `def upsample` I've added this change there

Comment (tapspatel) [reply tests/ttnn/unit_tests/operations/test_silu.py:26]: alternatively, you can also use device object `device.arch() == ttl.device.Arch.WORMHOLE_B0`

**Comment by shwetankTT** [comment]
@arakhmati @eyonland Can you please take a look at this PR.

**Review by eyonland — APPROVED** [review]

File: tt_eager/tt_dnn/op_library/eltwise_unary/multi_core/eltwise_unary_op_multi_core_height_block_sharded.cpp [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+                    core_coord,
+                    {
+                        (uint32_t)(num_tile_per_core)
+                    }
+                );
+           }
+        }
+    }else{
+        TT_FATAL(false, "Only height and block memory is supported by this fuction");
+    }
```
Comment (tt-aho) [inline tt_eager/tt_dnn/op_library/eltwise_unary/multi_core/eltwise_unary_op_multi_core_height_block_sharded.cpp:149]: There is no need for all this code/loops if you are just supporting sharded->sharded?
Can't it just be replaced with

```
std::vector<uint32_t> reader_runtime_args = {(uint32_t)(num_tile_per_core)};
tt_metal::SetRuntimeArgs(program, unary_reader_kernel_id, all_cores, reader_runtime_args);
```

File: tt_eager/tt_dnn/op_library/eltwise_unary/eltwise_unary_op.cpp [hunk 11]
```diff
@@ -258,8 +258,6 @@ void EltwiseUnary::validate(const std::vector<Tensor> &input_tensors) const {
     TT_FATAL(input_tensor_a.storage_type() == StorageType::DEVICE, "Operands to eltwise unary need to be on device!");
     TT_FATAL(input_tensor_a.buffer() != nullptr , "Operands to eltwise unary need to be allocated in buffers on device!");
     TT_FATAL((input_tensor_a.get_layout() == Layout::TILE), "Inputs to eltwise unary must be tilized");
-    TT_FATAL(input_tensor_a.memory_config().memory_layout == TensorMemoryLayout::INTERLEAVED, "Eltwise unary does not currently support sharding");
-    TT_FATAL(this->output_mem_config.memory_layout == TensorMemoryLayout::INTERLEAVED, "Eltwise unary does not currently support sharding");
```
[the remark this answers is not in the document]
Comment (shwetankTT) [reply tt_eager/tt_dnn/op_library/eltwise_unary/eltwise_unary_op.cpp:262]: thanks. Will address these helpful feedbacks.

**Pull request merged**